domain/model/service/service_test.go

```diff
@@ -844,6 +845,66 @@ func (s *serviceSuite) TestControllerModel(c *gc.C) {
 	})
 }
 
+func (s *serviceSuite) TestGetModelUsersAll(c *gc.C) {
+	uuid, err := coremodel.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	bobName := usertesting.GenNewName(c, "bob")
+	jimName := usertesting.GenNewName(c, "jim")
+	adminName := usertesting.GenNewName(c, "admin")
+	s.state.users = map[user.UUID]user.Name{
```

Comment:
Maybe out of scope for this PR, but I would prefer to use the real state for these tests, we can't guarantee behavioural equivalence with this fake implementation.

Reply:
Imo this fake implementation is a misfeature and it should just be mocks. Like all dummy implementations you spend half the time just debugging it rather than the actual code.
Mocks wouldn't give guaranteed behavioral equivalence but at least they would make it very clear what the interface was expected to be. From what I understand, the idea of using mocks here rather than just the real state is to avoid getting into a situation like with mongo where changing something in state meant you had to change tests all over the place because they relied on implementation details. At least with mocks you only have to change the tests if the interface changes (though that includes the expected errors from the state which won't cause any compile errors if changed).

Reply:
We've run into problems mocking state in the service layer because often most of the functionality of the service layer is provided by the state layer (e.g. returned errors).

> the idea of using mocks here rather than just the real state is to avoid getting into a situation like with mongo where changing something in state meant you had to change tests all over the place because they relied on implementation details.

But then the tests *should* change. For example if we change the error types return by the state method, the interface of the service has changed, so we should have a corresponding test change.

I agree at the level of facades and above we should mock dependencies, the current system of spinning up an entire DB to test facades or even the client is horrible. Just in this particular case of services, I think integration tests are a better choice than mocking.

Anyway as I said, this might be out of scope of this PR.